core/src/main/java/com/linecorp/armeria/client/HttpResponseDecoder.java

```diff
@@ -253,84 +235,57 @@ public void abort(Throwable cause) {
-         * from {@link Http1ResponseDecoder} and {@link Http2ResponseDecoder}. If this returns {@code false},
-         * it means the response stream has been closed due to disconnection or by the response consumer.
-         * So the caller do not need to handle such cases because it will be notified to the response
-         * consumer anyway.
-         */
         @Override
         public boolean tryWrite(HttpObject o) {
-            boolean wrote = false;
-            switch (state) {
-                case WAIT_NON_INFORMATIONAL:
-                    wrote = handleWaitNonInformational(o);
-                    break;
-                case WAIT_DATA_OR_TRAILERS:
-                    wrote = handleWaitDataOrTrailers(o);
-                    break;
-                case DONE:
-                    PooledObjects.close(o);
-                    break;
+            if (done) {
+                PooledObjects.close(o);
+                return false;
             }
-
-            return wrote;
-        }
-
-        @Override
-        public boolean tryWrite(Supplier<? extends HttpObject> o) {
             return delegate.tryWrite(o);
         }
 
-        private boolean handleWaitNonInformational(HttpObject o) {
-            // NB: It's safe to call logBuilder.startResponse() multiple times.
-            if (ctx != null) {
-                ctx.logBuilder().startResponse();
+        void startResponse() {
+            if (responseStarted) {
+                return;
             }
-
-            assert o instanceof HttpHeaders && !(o instanceof RequestHeaders) : o;
-
-            if (o instanceof ResponseHeaders) {
-                final ResponseHeaders headers = (ResponseHeaders) o;
-                final HttpStatus status = headers.status();
-                if (!status.isInformational()) {
-                    this.headers = headers;
-                    state = State.WAIT_DATA_OR_TRAILERS;
-                    if (ctx != null) {
-                        ctx.logBuilder().defer(RequestLogProperty.RESPONSE_HEADERS);
-                        try {
-                            return delegate.tryWrite(headers);
-                        } finally {
-                            ctx.logBuilder().responseHeaders(headers);
-                        }
-                    }
-                }
+            responseStarted = true;
+            ctx.logBuilder().startResponse();
+            ctx.logBuilder().responseFirstBytesTransferred();
+            initTimeout();
+        }
+
+        boolean tryWriteResponseHeaders(ResponseHeaders responseHeaders) {
+            assert responseHeaders.status().codeClass() != HttpStatusClass.INFORMATIONAL;
+            contentLengthHeaderValue = responseHeaders.contentLength();
+            ctx.logBuilder().defer(RequestLogProperty.RESPONSE_HEADERS);
+            try {
+                return delegate.tryWrite(responseHeaders);
+            } finally {
+                ctx.logBuilder().responseHeaders(responseHeaders);
```

Comment:
Q: Why do we need to defer `RESPONSE_HEADERS`? It is set immediately. Is there a timing issue?

Reply:
I believe that there's a chance that `endResponse` is called by the `tryWrite` so we needed to defer it. 
https://github.com/line/armeria/pull/2832

Reply:
I see. Thanks for the link.